include/swift-c/DependencyScan/DependencyScan.h

```diff
@@ -492,6 +538,151 @@ SWIFTSCAN_PUBLIC swiftscan_string_ref_t
+    swiftscan_cas_t cas, swiftscan_cached_output_t, void *ctx,
+    void (*callback)(void *ctx, bool success, swiftscan_string_ref_t error),
+    swiftscan_cache_cancellation_token_t *);
+
+/// Check if cached output is materialized locally and can be accessed
+/// without downloading.
+SWIFTSCAN_PUBLIC bool
+    swiftscan_cached_output_is_materialized(swiftscan_cas_t,
+                                            swiftscan_cached_output_t);
+
+/// Dispose a cached output.
+SWIFTSCAN_PUBLIC
+void swiftscan_cached_output_dispose(swiftscan_cached_output_t);
+
+/// Cancel the async cache action that is associated with token.
+SWIFTSCAN_PUBLIC void
+    swiftscan_cache_action_cancel(swiftscan_cache_cancellation_token_t);
+
+/// Dispose the cancellation token.
+SWIFTSCAN_PUBLIC void swiftscan_cache_cancellation_token_dispose(
+    swiftscan_cache_cancellation_token_t);
+
+/// Create a swift cached compilation replay instance with its command-line
+/// invocation. Return nullptr when errors occurs and the error message is
+/// returned via \c error parameter and its caller needs to free the message
+/// using `swiftscan_string_dispose`.
+SWIFTSCAN_PUBLIC swiftscan_cache_replay_instance_t
+swiftscan_cache_replay_instance_create(int argc, const char **argv,
+                                       swiftscan_string_ref_t *error);
+
+/// Dispose swift cached compilation replay instance.
+SWIFTSCAN_PUBLIC void
+    swiftscan_cache_replay_instance_dispose(swiftscan_cache_replay_instance_t);
+
+/// Replay the cached compilation using cached compliation replay instance.
+/// Returns replay result or nullptr if output not found or error occurs. If
+/// error, the error message is returned via \c error parameter and its caller
+/// needs to free the message using `swiftscan_string_dispose`.
+SWIFTSCAN_PUBLIC swiftscan_cache_replay_result_t
+swiftscan_cache_replay_compilation(swiftscan_cas_t cas,
+                                   swiftscan_cache_replay_instance_t,
+                                   swiftscan_cached_compilation_t,
+                                   swiftscan_string_ref_t *error);
+
+/// Async replay the cached compilation. Callback function will be provided
+/// by the cache replay result, and its stdout, stderr, and any error message.
+/// \c ctx is an opaque value that passed to the callback and
+/// \c swiftscan_cache_cancellation_token_t will return an token that can be
+/// used to cancel the async operation. The token needs to be freed by caller
+/// using `swiftscan_cache_cancellation_token_dispose`.
+SWIFTSCAN_PUBLIC void swiftscan_cache_replay_compilation_async(
+    swiftscan_cas_t cas, swiftscan_cache_replay_instance_t,
+    swiftscan_cached_compilation_t, void *ctx,
+    void (*callback)(void *ctx, swiftscan_cache_replay_result_t,
+                     swiftscan_string_ref_t error),
+    swiftscan_cache_cancellation_token_t *);
+
+/// Get return code from replay. Return code is 0 for success, and a non-zero
+/// value for failure.
```

Comment:
To clarify, a "non-zero value for failure" means we cached a compilation that fails when replayed? Or does it mean that the replay failed for some reason? If it is the latter shouldn't it fail at the `swiftscan_cache_replay_compilation` call?

Reply:
My thinking is returnCode should be whatever the `swift-frontend` invocation should return. It is hard to match exactly but maybe I should document it to be like that?

Reply:
So a "non-zero value" means we cached a compilation that fails when replayed? Which in practice will not happen because we won't cache a failed invocation?
I'd suggest to make it clear in the doc-comment this is the cached result and non-zero doesn't mean the replay failed.

Reply:
No, non-zero value absolutely means the replay failed. The returnCode matches the swift-frontend's return code if it is used for cache replay. Let me clarify that.

Reply:
So if the replay failed where is the client getting the error description from, the `error` parameter of `swiftscan_cache_replay_compilation` or `swiftscan_cache_replay_result_get_stderr`?

Reply:
There is intentionally no error parameter. The replay is behaving like a process. It has a return code and it has stdout/error. It can be counted to have a nice diagnostics from stderr, but it is harder to implement a solution that the error message has to comeback from the parameter.

Reply:
Oh, I see, I actually have an error parameter here. Hmm... let me think about it.

Reply:
I see, with current implementation, the returnCode is not that useful. Let me get rid of it and using error parameter to indicate if there are any error.